readthedocs/core/resolver.py

```diff
@@ -53,65 +54,79 @@ class ResolverBase:
     """
 
     def base_resolve_path(
-            self,
-            project_slug,
-            filename,
-            version_slug=None,
-            language=None,
-            single_version=None,
-            subproject_slug=None,
-            subdomain=None,
-            cname=None,
-            urlconf=None,
+        self,
+        project_slug,
+        filename,
+        version_slug=None,
+        language=None,
+        single_version=None,
+        subproject_relationship=None,
+        subdomain=None,
+        cname=None,
+        urlconf=None,
+        urlpattern=None,
     ):
         """Resolve a with nothing smart, just filling in the blanks."""
         # Only support `/docs/project' URLs outside our normal environment. Normally
         # the path should always have a subdomain or CNAME domain
-        if subdomain or cname or (self._use_subdomain()):
-            url = '/'
+        if subdomain or cname or self._use_subdomain():
+            path = "/"
         else:
-            url = '/docs/{project_slug}/'
+            path = "/docs/{project}/"
 
-        if subproject_slug:
-            url += 'projects/{subproject_slug}/'
+        if subproject_relationship:
+            path = unsafe_join_url_path(
+                path, subproject_relationship.child.subproject_prefix
+            )
 
-        if single_version:
-            url += '{filename}'
+        # If the project has a custom urlpattern, we use it.
+        # If the project doesn't have a custom URL pattern, we use the default,
+        # which is just the filename for single version projects and language/version
+        # for multi version projects.
+        if urlpattern:
+            plain_urlpattern = urlpattern_to_plain_text(urlpattern)
+            path = unsafe_join_url_path(path, plain_urlpattern)
+        elif single_version:
+            path = unsafe_join_url_path(path, "{filename}")
         else:
-            url += '{language}/{version_slug}/{filename}'
+            path = unsafe_join_url_path(path, "{language}/{version}/{filename}")
 
+        # TODO: remove this when all projects have migrated to URL patterns.
         # Allow users to override their own URLConf
         # This logic could be cleaned up with a standard set of variable replacements
         if urlconf:
-            url = urlconf
-            url = url.replace(
-                '$version',
-                '{version_slug}',
+            path = urlconf
+            path = path.replace(
+                "$version",
+                "{version}",
```

Comment:
Do we want to convert these, instead of just defining the user-facing value with a `{version}` syntax? That seems simpler and perhaps removes the hacks needed for unescaping the string?

Reply:
This is part of the old urlconf implementation, so I don't think we should complicate the deploy for something that will be deleted soon.

Reply:
Gotcha 👍